utils/yoke/src/yoke.rs

```diff
@@ -480,50 +471,183 @@ where
+    /// produce another Yokeable (`P`) that also borrows from the same cart. There are a couple potential
+    /// hazards here:
+    ///
+    /// - `P` ends up borrowing data from `Y` (or elsewhere) that did _not_ come from the cart,
+    ///   for example `P` could borrow owned data from a `Cow`.
+    /// - Borrowed data from `Y` escapes with the wrong lifetime
+    ///
+    /// Let's walk through these and see how they're prevented.
+    ///
+    /// ```rust, compile_fail
+    /// # use std::rc::Rc;
+    /// # use yoke::Yoke;
+    /// # use std::borrow::Cow;
+    /// fn borrow_potentially_owned(y: Yoke<Cow<'static, str>, Rc<[u8]>>) -> Yoke<&'static str, Rc<[u8]>> {
+    ///    y.project(|cow, _| &*cow)   
+    /// }
+    /// ```
+    ///
+    /// In this case, the lifetime of `&*cow` is `&'this str`, however the function needs to be able to return
+    /// `&'a str` _for all `'a`_, which isn't possible.
+    ///
+    /// Similarly, trying to project an owned field of a struct will produce similar errors:
+    ///
+    /// ```rust,compile_fail
+    /// # use std::borrow::Cow;
+    /// # use yoke::{Yoke, Yokeable};
+    /// # use std::mem;
+    /// # use std::rc::Rc;
+    /// #
+    /// // also safely implements Yokeable<'a>
+    /// struct Bar<'a> {
+    ///     owned: String,
+    ///     string_2: &'a str,
+    /// }
+    ///
+    /// fn project_owned(bar: &Yoke<Bar<'static>, Rc<[u8]>>) -> Yoke<&'static str, Rc<[u8]>> {
+    ///     // ERROR (but works if you replace owned with string_2)
+    ///     bar.project(|bar, _| &*bar.owned)   
+    /// }
+    ///
+    /// #
+    /// # unsafe impl<'a> Yokeable<'a> for Bar<'static> {
+    /// #     type Output = Bar<'a>;
+    /// #     fn transform(&'a self) -> &'a Bar<'a> {
+    /// #         self
+    /// #     }
+    /// #
+    /// #     unsafe fn make(from: Bar<'a>) -> Self {
+    /// #         let ret = mem::transmute_copy(&from);
+    /// #         mem::forget(from);
+    /// #         ret
+    /// #     }
+    /// #
+    /// #     fn transform_mut<F>(&'a mut self, f: F)
+    /// #     where
+    /// #         F: 'static + FnOnce(&'a mut Self::Output),
+    /// #     {
+    /// #         unsafe { f(mem::transmute(self)) }
+    /// #     }
+    /// # }
+    /// ```
+    ///
+    /// Borrowed data from `Y` similarly cannot escape with the wrong lifetime because of the `for<'a>`, since
+    /// it will never be valid for the borrowed data to escape for all lifetimes of 'a. Internally, `.project()`
+    /// uses `.get()`, however the signature forces the callers to be able to handle every lifetime.
+    ///
+    /// Essentially, safety is achieved by using `for<'a> fn(...)` with `'a` used in both `Yokeable`s to ensure that
+    /// the output yokeable can _only_ have borrowed data flow in to it from the input, and the `'this` ensures separation
+    /// between `'a` data that the input is borrowing vs data that the input owns.
```

Comment:
Indeed, and the reason this separation works is because at the symbolic level, existential lifetimes and universal lifetimes can never be equal for all lifetimes ('this is basically existential from the point of view of the closure).